packages/steps/src/BuildConfig.ts

```diff
@@ -71,32 +72,54 @@ export type BuildFunctionInputs = (
   | string
   | {
       name: string;
-      defaultValue?: string | boolean;
-      allowedValues?: (string | boolean)[];
+      defaultValue?: BuildStepInputValueType;
+      allowedValues?: BuildStepInputValueType[];
       required?: boolean;
+      allowedValueType: BuildStepInputValueTypeName;
     }
 )[];
 export type BuildFunctionOutputs = BuildStepOutputs;
 
 const BuildFunctionInputsSchema = Joi.array().items(
-  Joi.alternatives().try(
-    Joi.string().required(),
-    Joi.object({
+  Joi.alternatives().conditional(Joi.ref('.'), {
+    is: Joi.string(),
+    then: Joi.string().required(),
+    otherwise: Joi.object({
       name: Joi.string().required(),
-      defaultValue: Joi.alternatives().conditional('allowedValues', {
+      defaultValue: Joi.when('allowedValues', {
         is: Joi.exist(),
         then: Joi.valid(Joi.in('allowedValues')).messages({
           'any.only': '{{#label}} must be one of allowed values',
         }),
-        otherwise: Joi.alternatives().try(Joi.string(), Joi.boolean()),
+        otherwise: Joi.when('allowedValueType', {
+          is: BuildStepInputValueTypeName.STRING,
+          then: Joi.string(),
+          otherwise: Joi.when('allowedValueType', {
+            is: BuildStepInputValueTypeName.BOOLEAN,
+            then: Joi.boolean(),
+            otherwise: Joi.number(),
+          }),
+        }),
+      }),
+      allowedValues: Joi.when('allowedValueType', {
+        is: BuildStepInputValueTypeName.STRING,
+        then: Joi.array().items(Joi.string()),
+        otherwise: Joi.when('allowedValueType', {
+          is: BuildStepInputValueTypeName.BOOLEAN,
+          then: Joi.array().items(Joi.boolean()),
+          otherwise: Joi.array().items(Joi.number()),
+        }),
```

Comment:
```suggestion
      allowedValues: Joi.when('allowedValueType', {
        is: BuildStepInputValueTypeName.STRING,
        then: Joi.array().items(Joi.string()),
      })
      .when('allowedValueType', {
          is: BuildStepInputValueTypeName.BOOLEAN,
          then: Joi.array().items(Joi.boolean()),
      }).
      .when('allowedValueType', {
          is: BuildStepInputValueTypeName.NUMBER,
          then: Joi.array().items(Joi.boolean()),
       }),
```

Wouldn't sth like this work? (not tested)